### PR TITLE
Add nextcloud integration

### PR DESCRIPTION
**Pull request by cytadela8**

### CODE REVIEW

File: wwwapp/views.py [hunk 1]
```diff
@@ -793,3 +807,32 @@ def upload_file(request, type, name):
...
+    if userprofile.owncloud_user == "":
+        userprofile.owncloud_user = \
+            unicodedata.normalize("NFKD", request.user.first_name+request.user.last_name)\
+                .encode('ascii','ignore').decode('ascii')[:32]
+        userprofile.owncloud_password = generate_nice_pass()
+        c = owncloud.Client(settings.OWNCLOUD_HOST)
+        c.login(settings.OWNCLOUD_USER, settings.OWNCLOUD_PASS)
+        c.create_user(userprofile.owncloud_user, userprofile.owncloud_password)
+        if is_lecturer:
+            c.add_user_to_group(userprofile.owncloud_user, "WWW16Lecturers")
```
Comment (krzys-h) [inline wwwapp/views.py:830]: This should probably be configurable as well, but we can address this later (add an issue sowe wo't forget)

File: templates/base.html [hunk 2]
```diff
@@ -84,6 +84,10 @@
               </li>
             {% endif %}
 
+            {% if is_participating_this_year %}
+              <li><a href="{% url 'cloud_access' %}">Chmura</a></li>
```
Comment (krzys-h) [inline templates/base.html:88]: This should be available only if owncloud is actually configured, but we can fix this later too

File: wwwapp/views.py [hunk 3]
```diff
@@ -793,3 +807,32 @@ def upload_file(request, type, name):
...
+        elif is_participant:
+            c.add_user_to_group(userprofile.owncloud_user, "WWW16Participants")
+        userprofile.save()
+
+    context['owncloud_host'] = settings.OWNCLOUD_HOST
```
Comment (krzys-h) [inline wwwapp/views.py:835]: Add sample config to settings files in the repo

File: wwwapp/views.py [hunk 4]
```diff
@@ -793,3 +807,32 @@ def upload_file(request, type, name):
...
+def cloud_access_view(request):
+    context = get_context(request)
+    userprofile = request.user.userprofile
+    is_lecturer = userprofile.is_lecturer_in(settings.CURRENT_YEAR)
+    is_participant = userprofile.is_participant_in(settings.CURRENT_YEAR)
+    if not (is_lecturer or is_participant):
+        return HttpResponseForbidden()
+
+    if userprofile.owncloud_user == "":
+        userprofile.owncloud_user = \
```
[the remark this answers is not in the document]
Comment (cytadela8) [reply wwwapp/views.py:822]: If it's not unique Owncloud will give as an error

File: wwwapp/models.py [hunk 5]
```diff
@@ -23,9 +23,17 @@ class UserProfile(models.Model):
     how_do_you_know_about = models.CharField(max_length=1000, default="", blank=True)
     profile_page = models.TextField(max_length=100000, blank=True, default="")
     cover_letter = models.TextField(max_length=100000, blank=True, default="")
+    owncloud_user = models.CharField(max_length=32, blank=True, default="")
```
[the remark this answers is not in the document]
Comment (cytadela8) [reply wwwapp/models.py:26]: meh

File: wwwapp/views.py (hunk 1 above)
Comment (cytadela8) [reply wwwapp/views.py:830]: Who cares? This code is only for this year.

**Pull request merged**